packaging/os/yumrepo.py

```diff
@@ -549,40 +914,81 @@ def main():
     name = module.params['name']
     state = module.params['state']
 
-    # Rename "name" and "description" to ensure correct key sorting
-    module.params['repoid'] = module.params['name']
-    module.params['name'] = module.params['description']
-    del module.params['description']
-
-    # Instantiate the YumRepo object
-    yumrepo = YumRepo(module)
-
-    # Get repo status before change
-    yumrepo_before = yumrepo.dump()
-
-    # Perform action depending on the state
+    # Check if required parameters are present
+    if state in ['present', 'absent'] and 'name' is None:
+        module.fail_json(
+            msg="Parameter 'name' is required.")
     if state == 'present':
-        yumrepo.add()
-    elif state == 'absent':
-        yumrepo.remove()
-
-    # Get repo status after change
-    yumrepo_after = yumrepo.dump()
-
-    # Compare repo states
-    changed = yumrepo_before != yumrepo_after
-
-    # Save the file only if not in check mode and if there was a change
-    if not module.check_mode and changed:
-        yumrepo.save()
-
-    # Change file attributes if needed
-    if os.path.isfile(module.params['dest']):
-        file_args = module.load_file_common_arguments(module.params)
-        changed = module.set_fs_attributes_if_different(file_args, changed)
-
-    # Print status of the change
-    module.exit_json(changed=changed, repo=name, state=state)
+        if (
+                'baseurl' is None and
+                'mirrorlist' is None):
+            module.fail_json(
+                msg="Parameter 'baseurl' or 'mirrorlist' is required.")
+        if 'description' is None:
+            module.fail_json(
+                msg="Parameter 'description' is required.")
+
+    # Instantiate the ManagedMode object
+    mm = ManagedMode(module)
+
+    if state in ['present', 'absent']:
+        # Rename "name" and "description" to ensure correct key sorting
+        module.params['repoid'] = module.params['name']
+        module.params['name'] = module.params['description']
+        del module.params['description']
+        del module.params['ignore_repo_files']
+
+        # Define repo file name if it doesn't exist
+        if module.params['file'] is None:
+            module.params['file'] = module.params['repoid']
```

Comment:
It might be somewhat obvious, but this will fail if an already existing repo file has a section (`repoid`) which name is not the same as the file name. 

Works as described using the explicit `file` parameter, but the documentation could possibly be updated to include a "warning" of sorts for the `name: Remove repository` example. 

For example, the default CentOS-Media.repo in CentOS 7:

    $ cat CentOS-Media.repo                                                      
    # CentOS-Media.repo
    #
    #  This repo can be used with mounted DVD media, verify the mount point for
    #  CentOS-7.  You can use this repo and yum to install items directly off the
    #  DVD ISO that we release.
    #
    # To use this repo, put in your DVD and use it with the other repos too:
    #  yum --enablerepo=c7-media [command]
    #  
    # or for ONLY the media repo, do this:
    #
    #  yum --disablerepo=\* --enablerepo=c7-media [command]

    [c7-media]
    name=CentOS-$releasever - Media
    baseurl=file:///media/CentOS/
            file:///media/cdrom/
            file:///media/cdrecorder/
    gpgcheck=1
    enabled=0
    gpgkey=file:///etc/pki/rpm-gpg/RPM-GPG-KEY-CentOS-7
                                                                                 
Will fail using:

    - name: Remove CentOS-Media
      yumrepo:
        name: c7-media
        state: absent